sweepbatcher/sweep_batcher.go

```diff
@@ -0,0 +1,657 @@
+	}
+
+	batchKit := batchKit{
+		id:               batch.id,
+		batchTxid:        batch.batchTxid,
+		batchPkScript:    batch.batchPkScript,
+		state:            batch.state,
+		primaryID:        primarySweep.SwapHash,
+		sweeps:           sweeps,
+		rbfCache:         rbfCache,
+		returnChan:       b.sweepReqs,
+		wallet:           b.wallet,
+		chainNotifier:    b.chainNotifier,
+		signerClient:     b.signerClient,
+		musig2SignSweep:  b.musig2ServerSign,
+		verifySchnorrSig: b.VerifySchnorrSig,
+		purger:           b.AddSweep,
+		store:            b.store,
+		log:              batchPrefixLogger(fmt.Sprintf("%d", batch.id)),
+	}
+
+	newBatch := NewBatchFromDB(cfg, batchKit)
+
+	// We add the batch to our map of batches and start it.
+	b.batches[batch.id] = newBatch
+
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+
+		err := newBatch.Run(ctx)
+		if err != nil {
+			_ = b.writeToErrChan(ctx, err)
+		}
+	}()
+
+	return nil
+}
+
+// FetchUnconfirmedBatches fetches all the batches from the database that are
+// not in a confirmed state.
+func (b *Batcher) FetchUnconfirmedBatches(ctx context.Context) ([]*batch,
+	error) {
+
+	dbBatches, err := b.store.FetchUnconfirmedSweepBatches(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	batches := make([]*batch, 0, len(dbBatches))
+	for _, bch := range dbBatches {
+		bch := bch
+
+		batch := batch{}
+		batch.id = bch.ID
+
+		switch bch.State {
+		case batchOpen:
+			batch.state = Open
+
+		case batchClosed:
+			batch.state = Closed
+
+		case batchConfirmed:
+			batch.state = Confirmed
+		}
+
+		batch.batchTxid = &bch.BatchTxid
+		batch.batchPkScript = bch.BatchPkScript
+
+		rbfCache := rbfCache{
+			LastHeight: bch.LastRbfHeight,
+			FeeRate:    chainfee.SatPerKWeight(bch.LastRbfSatPerKw),
+		}
+		batch.rbfCache = rbfCache
+
+		bchCfg := batchConfig{
+			maxTimeoutDistance: bch.MaxTimeoutDistance,
+		}
+		batch.cfg = &bchCfg
+
+		batches = append(batches, &batch)
+	}
+
+	return batches, nil
+}
+
+// monitorSpendAndNotify monitors the spend of a specific outpoint and writes
+// the response back to the response channel.
+func (b *Batcher) monitorSpendAndNotify(ctx context.Context, sweep *sweep,
+	notifier *SpendNotifier) {
+
+	b.wg.Add(1)
+	defer b.wg.Done()
+
+	spendCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	spendChan, spendErr, err := b.chainNotifier.RegisterSpendNtfn(
+		spendCtx, &sweep.outpoint, sweep.htlc.PkScript,
+		sweep.initiationHeight,
+	)
+	if err != nil {
+		select {
+		case notifier.SpendErrChan <- err:
+		case <-ctx.Done():
+		}
+
+		_ = b.writeToErrChan(ctx, err)
+
+		return
+	}
+
+	log.Infof("Batcher monitoring spend for swap %x", sweep.swapHash[:6])
+
+	for {
+		select {
+		case spend := <-spendChan:
+			select {
+			case notifier.SpendChan <- spend.SpendingTx:
+			case <-ctx.Done():
+			}
+
+			return
+
+		case err := <-spendErr:
+			select {
+			case notifier.SpendErrChan <- err:
+			case <-ctx.Done():
+			}
+
+			_ = b.writeToErrChan(ctx, err)
```

Comment:
q: Do we need to add this error to the batcher's error channel? It'll stop the batcher altogether.

Reply:
Hmm, not sure. If we don't it will only fail on the batch/swap level. One reason to error out the whole daemon is that a restart **may** fix the monitor spend issue. But that could also "deadlock" the whole daemon just for the sake of 1 swap